tripper/dataset/dataset.py

```diff
@@ -0,0 +1,942 @@
+    else:
+        raise TypeError(
+            "if given, `dataset` should be either a string or dict"
+        )
+    dataset: dict  # Tell mypy that this now is a dict
+
+    if distribution is None:
+        if "distribution" in dataset:
+            distribution = get(dataset, "distribution")[0]
+        else:
+            newiri = f"_:N{secrets.token_hex(16)}"
+            distribution = AttrDict(
+                {"@id": newiri, "@type": DCAT.Distribution}
+            )
+            add(dataset, "distribution", distribution)
+            triples.append((dataset["@id"], DCAT.distribution, newiri))
+            save_distribution = True
+    if isinstance(distribution, str):
+        distr = load_dict(ts, iri=distribution, use_sparql=use_sparql)
+        if distr:
+            distribution = distr
+        else:
+            distribution = AttrDict(
+                {"@id": distribution, "@type": DCAT.Distribution}
+            )
+            add(dataset, "distribution", distribution)
+            triples.append((dataset["@id"], DCAT.distribution, newiri))
+            save_distribution = True
+    elif isinstance(distribution, dict):
+        add(dataset, DCAT.distribution, distribution)
+        if "@id" in distribution:
+            triples.append(
+                (dataset["@id"], DCAT.distribution, distribution["@id"])
+            )
+        save_distribution = True
+    else:
+        raise TypeError(
+            "if given, `distribution` should be either a string or dict"
+        )
+    distribution: dict  # Tell mypy that this now is a dict
+
+    if isinstance(generator, str):
+        for gen in get(distribution, "generator"):
+            if gen.get("@id") == generator:
+                break
+        else:
+            raise ValueError(
+                f"dataset '{dataset}' has no such generator: {generator}"
+            )
+    elif "generator" in distribution:
+        gen = get(distribution, "generator")[0]
+    else:
+        gen = None
+
+    # __TODO__: Check if `class_iri` already has the value restriction.
+    # If not, add it to triples
+
+    # __TODO__: Move this mapping of supported schemes into the protocol
+    # module.
+    schemes = {
+        "https": "http",
+    }
+
+    # Save data
+    url = distribution.get("downloadURL", distribution.get("accessURL"))
+    p = urlparse(url)
+    scheme = schemes.get(p.scheme, p.scheme) if p.scheme else "file"
+    location = (
+        f"{scheme}://{p.netloc}{p.path}" if p.netloc else f"{scheme}:{p.path}"
+    )
+    options = [p.query] if p.query else []
+    if gen and "configuration" in gen and "options" in gen.configuration:
+        # __TODO__: allow options to also be a dict
+        options.append(gen.configuration["options"])
+    id = (
+        distribution["accessService"].get("identifier")
+        if "accessService" in distribution
+        else None
+    )
+    with Protocol(scheme, location, options=";".join(options)) as pr:
+        pr.save(data, id)
```

Comment:
Seems strange that ther are not other, more persistent packages we can use for this.